### ISSUE TITLE
how to correctly parse domains which are listed within PSL

### ISSUE THREAD

**Issue by deepika-maj** [jeremykendall/php-domain-parser#251]
I am trying to parse domains to extract information like subdomain, registrable domain and public suffix or TLD. Everything was working fine until I encounter domains which are listed within the PSL.

I also noticed two issues #240 and #244 which seems to be related to mine, but I could not figure out how to solve this so that I can correctly extract information.

For an example, below are the list based on real world scenario:

SN | Input Domain | Sub Domain | Registrable Domain | Public Suffix or TLD | Parsing Correctly
-- | - | - | - | - | -
1 | google.com | null | google.com | com | true
2 | docs.google.com | docs | google.com | com | true
3 | sa.gov.au | null | sa.gov.au | gov.au | false
4 | kvam.no | null | kvam.no | no | false
5 | unb.com.bd | null | unb.com.bd | com.bd | true

and the code to test it:
```php
<?php

use Pdp\Cache;
use Pdp\CurlHttpClient;
use Pdp\Manager;
use Pdp\Rules;

$pslManager = new Manager(new Cache(), new CurlHttpClient());
$pslRules = $pslManager->getRules()
    ->withAsciiIDNAOption(IDNA_NONTRANSITIONAL_TO_ASCII)
    ->withUnicodeIDNAOption(IDNA_NONTRANSITIONAL_TO_UNICODE);

$hostNames = [
    'google.com',
    'docs.google.com',
    'sa.gov.au',
    'kvam.no',
    'unb.com.bd'
];

foreach ($hostNames as $hostName) {
    $resp = $pslRules->resolve($hostName, Rules::ICANN_DOMAINS)->toAscii();
    print_r($resp->jsonSerialize());
}
```

I also checked the  "Algorithm" and "Definitions" listed at https://publicsuffix.org/list/ and based on it, the library should have parsed the data correctly.

> When the domain and rule are split into corresponding labels, that the domain contains as many or more labels than the rule.

> The registered or registrable domain is the public suffix plus one additional label.

could someone guide me into correct direction

**Comment by nyamsprod** [comment]
@deepika-maj are you sure that `sa.gov.au` and `kvam.no` are not listed on [the PSL ](https://publicsuffix.org/list/public_suffix_list.dat)? Because if they are listed there in the ICANN section then the responses you are getting are expected.

**Comment by deepika-maj** [comment]
@nyamsprod yes both `sa.gov.au` and `kvam.no` exists under ICANN section, however as mentioned in #240 it can not be parsed/resolved by using:

`$pslRules->resolve('sa.gov.au', Rules::ICANN_DOMAINS);`

**Comment by nyamsprod** [comment]
Which is the expected behavior as already explained twice in the issues and in the pinned issue.

**Comment by deepika-maj** [comment]
so, does this means such domains can not be parsed using this library?

**Comment by nyamsprod** [comment]
**TL;DR:** a domain can be parsable **WHILE** not being resolvable these are two separate properties. 

A **parsable domain** is one that you can create using the following code:

```php
$domain = new Domain('sa.gov.au'); //this works
```

A **resolvable domain**  is a domain which has at least a registrable part.

```php
$rules->resolve('foo.sa.gov.au');
```
is resolvable and its public suffix part will is `sa.gov.au`

```php
$rules->resolve('sa.gov.au');
```
 is not resolvable because `sa.gov.au` is a public suffix according to the ICANN section.

**Comment by deepika-maj** [comment]
I am still at confusion on how to make it work.

```php
$domain = new Domain('sa.gov.au');
```

The above code mostly creates an empty object:

```php
array (
  'domain' => 'sa.gov.au',
  'registrableDomain' => NULL,
  'subDomain' => NULL,
  'publicSuffix' => NULL,
  'isKnown' => false,
  'isICANN' => false,
  'isPrivate' => false,
)
```

Since this is a domain parser library, if the passed domain (i.e. `sa.gov.au`) is already in PSL, why it does not ignore it and detect `gov.au` as a suffix returning `sa.gov.au` as a registrable domain with other keys and values.

**Comment by nyamsprod** [comment]
PSL does not contains domains but Public Suffix information which are two different things. I understand your confusion but I don't think a issue tracker is the best place to answer this type of question.

**Comment by gsouf** [comment]
@nyamsprod same issue with `sa.gov.au` the library return `isknown = false` that makes it not possible to validate it has a valid domain, when it is actually valid.
